File: labview/src/auth/throttle.ts
/**
 * Failed-attempt throttling for the login route.
 *
 * **Keyed by username, not by source address.** LabView is meant to sit behind a
 * tunnel and a reverse proxy, where every request arrives from the same handful of
 * addresses — the proxy's. Keying on the peer would put the whole fleet in one bucket,
 * so six wrong passwords from one browser would lock everybody out, and keying on
 * `X-Forwarded-For` would key on a header anyone can set. The username is the thing
 * being guessed, so it is the thing counted.
 *
 * The consequence is deliberate and worth stating: an attacker who knows a username
 * can keep that one account locked out. For a read-only dashboard with a handful of
 * operators that is a far better failure than either alternative, and the operator can
 * still reach `/api/healthz`, the logs and the container.
 *
 * `now` is a parameter, never read from the clock, because every rule here is a
 * comparison against it.
 */

/** What the login route should do about an attempt. */
export interface ThrottleDecision {
  allowed: boolean;
  /** Seconds to wait, for the `Retry-After` header. `0` when allowed. */
  retryAfterSeconds: number;
}

interface Bucket {
  failures: number;
  /** Epoch ms of the most recent failure — both the window anchor and the eviction key. */
  last: number;
}

export class LoginThrottle {
  private readonly buckets = new Map<string, Bucket>();

  /**
   * @param maxFailures attempts allowed within the window before locking.
   * @param lockoutSeconds how long a lock lasts, and how much quiet resets the count.
   * @param max distinct usernames tracked at once.
   */
  constructor(
    private readonly maxFailures: number,
    private readonly lockoutSeconds: number,
    private readonly max = 4096,
  ) {}

  /**
   * The key a username counts against.
   *
   * Lower-cased, so `Alice` and `alice` share a budget — otherwise case alone
   * multiplies the allowance by every capitalisation of the name. Callers pass the
   * output of `sanitizeUsername`, which collapses every unusable name to `"?"`: all
   * junk usernames then share one bucket, which is the right answer, since a script
   * spraying random names is one attacker.
   */
  static key(sanitizedUser: string): string {
    return sanitizedUser.toLowerCase();
  }

  /**
   * Whether an attempt on `key` may proceed.
   *
   * Read-only — asking does not consume an attempt, so this can be called before the
   * password is verified without a correct password ever being charged for one.
   */
  check(key: string, now: Date): ThrottleDecision {
    const b = this.buckets.get(key);
    if (!b || b.failures < this.maxFailures) return { allowed: true, retryAfterSeconds: 0 };
    const elapsed = (now.getTime() - b.last) / 1000;
    const remaining = this.lockoutSeconds - elapsed;
    if (remaining <= 0) {
      // The window closed. Drop the bucket rather than zeroing it, so an abandoned
      // name stops occupying a slot.
      this.buckets.delete(key);
      return { allowed: true, retryAfterSeconds: 0 };
    }
    return { allowed: false, retryAfterSeconds: Math.max(1, Math.ceil(remaining)) };
  }

  /**
   * Charge a failure.
   *
   * The window is anchored to the *latest* failure, not the first, so the count only
   * resets after `lockoutSeconds` of quiet. Anchoring to the first would let an
   * attacker pace their guesses to the window boundary and never be locked out.
   */
  fail(key: string, now: Date): ThrottleDecision {
    const t = now.getTime();
    const existing = this.buckets.get(key);
    const stale = existing !== undefined && (t - existing.last) / 1000 > this.lockoutSeconds;
    if (existing && !stale) {
      existing.failures += 1;
      existing.last = t;
    } else {
      this.evictIfFull();
      this.buckets.set(key, { failures: 1, last: t });
    }
    return this.check(key, now);
  }

  /** Clear the count for `key`. A correct password ends the lockout. */
  succeed(key: string): void {
    this.buckets.delete(key);
  }

  /** Drop every bucket whose window has closed. */
  prune(now: Date): void {
    const cutoff = now.getTime() - this.lockoutSeconds * 1000;
    for (const [k, b] of this.buckets) if (b.last <= cutoff) this.buckets.delete(k);
  }

  get size(): number {
    return this.buckets.size;
  }

  /**
   * Make room at the cap by dropping the least recently failed bucket.
   *
   * Bounded because the key comes from a request body: without a cap, a script posting
   * a million distinct usernames would be a memory leak rather than a lockout. The
   * oldest entry is the one closest to expiring anyway, so evicting it is the smallest
   * loss — and losing a bucket loses a lockout, not a credential.
   */
  private evictIfFull(): void {
    if (this.buckets.size < this.max) return;
    let oldest: string | undefined;
    let oldestAt = Infinity;
    for (const [k, b] of this.buckets) {
      if (b.last < oldestAt) {
        oldest = k;
        oldestAt = b.last;
      }
    }
    if (oldest !== undefined) this.buckets.delete(oldest);
  }
}

File: labview/internal/access/throttle.go
package access

import (
	"sync"
	"time"
)

// The login throttle (§19).
//
// **A locked-out name is refused regardless of whether the password was right.** That is the point of
// the whole mechanism: a lock that opens for the correct password is not a lock, it is a slightly
// slower guessing game. It also means the throttle's answer is computed before any hash is compared,
// which is why the check comes first and the verification second.
//
// The counter is per name and not per address. An attacker guessing one account moves between
// addresses freely, and a lab behind a tunnel presents every request from the same address — so a
// per-address counter would both miss the attack and lock out the whole household when one person
// mistyped.

// MaxThrottleKeys is the number of usernames tracked at once (§19).
//
// A bound is needed because the key comes off the wire: without one, a script posting a fresh name per
// request would grow the table until the process died — a memory exhaustion delivered through the
// login form. 4096 is far more names than a homelab has and small enough to be free.
const MaxThrottleKeys = 4096

// Throttle counts recent failures per username.
//
// Safe for concurrent use. Every method takes the current time as a parameter so the expiry and the
// eviction are testable without waiting (I7).
type Throttle struct {
	// Max is the failures allowed inside Window before a name is locked. Zero or negative disables
	// the throttle, which is a configuration and not an error.
	Max int

	// Window is how long failures are remembered, and therefore how long a lock lasts.
	Window time.Duration

	mu sync.Mutex

	// entries is keyed on the case-folded sanitised name (ThrottleKey), so `Admin` and `admin` are
	// one account as far as a lock is concerned and the table cannot be filled with arbitrary bytes.
	entries map[string]*attempts

	// order is insertion order, for the eviction §19 requires. A slice rather than a heap: 4096
	// entries evicted one at a time is a memmove of a few kilobytes, and a heap here would be more
	// code than the thing it saves.
	order []string
}

// attempts is one name's recent failures.
type attempts struct {
	// count is failures inside the window, and first is when the window opened. Storing the window's
	// start rather than each failure's time keeps this to two words — and it is what makes the lock
	// last exactly Window from the failure that triggered it rather than sliding forward with every
	// further attempt.
	count int
	first time.Time

	// last is when the most recent failure landed, used only for eviction ordering when the table is
	// full.
	last time.Time
}

// Verdict is what a throttle says about one attempt.
type Verdict struct {
	// Locked is whether this attempt must be refused without checking the password.
	Locked bool

	// RetryAfter is how long until the lock lifts, rounded up to a whole second because that is what
	// a Retry-After header carries. Always at least 1 second when locked: a header saying `0` invites
	// an immediate retry, which is the opposite of what was meant.
	RetryAfter time.Duration
}

// Allow reports whether an attempt for user may proceed.
//
// It counts nothing. A check that incremented would lock a name out for reading the login form, and
// would make the count depend on how many times the gate happened to consult it.
func (t *Throttle) Allow(user string, now time.Time) Verdict {
	if t.Max <= 0 {
		return Verdict{}
	}

	t.mu.Lock()
	defer t.mu.Unlock()

	e := t.entries[ThrottleKey(user)]
	if e == nil || t.expired(e, now) {
		return Verdict{}
	}
	if e.count < t.Max {
		return Verdict{}
	}

	remaining := t.Window - now.Sub(e.first)
	if remaining < time.Second {
		remaining = time.Second
	}
	return Verdict{Locked: true, RetryAfter: remaining.Round(time.Second)}
}

// Failed records a failed attempt and returns the verdict that now applies.
//
// Returning the new verdict, rather than requiring a second Allow, is what lets the gate answer the
// attempt that *reached* the limit with a 429 rather than telling that one caller their password was
// wrong and the next one that they are locked.
func (t *Throttle) Failed(user string, now time.Time) Verdict {
	if t.Max <= 0 {
		return Verdict{}
	}

	t.mu.Lock()
	defer t.mu.Unlock()

	key := ThrottleKey(user)
	e := t.entries[key]
	switch {
	case e == nil:
		if t.entries == nil {
			t.entries = map[string]*attempts{}
		}
		t.evict()
		e = &attempts{first: now}
		t.entries[key], t.order = e, append(t.order, key)
	case t.expired(e, now):
		// A window that closed starts a new one rather than continuing the old count. Otherwise two
		// mistyped passwords a month apart would combine into a lock-out.
		e.count, e.first = 0, now
	}

	e.count, e.last = e.count+1, now

	if e.count < t.Max {
		return Verdict{}
	}
	remaining := t.Window - now.Sub(e.first)
	if remaining < time.Second {
		remaining = time.Second
	}
	return Verdict{Locked: true, RetryAfter: remaining.Round(time.Second)}
}

// Succeeded clears a name's count (§19).
//
// **The reset is on success only.** A name whose owner has just proved who they are is not a name
// under attack, so continuing to count against it would lock out the person who got in — and the
// count that was accumulating was, by construction, somebody failing to be them.
func (t *Throttle) Succeeded(user string) {
	t.mu.Lock()
	defer t.mu.Unlock()

	key := ThrottleKey(user)
	if _, ok := t.entries[key]; !ok {
		return
	}
	delete(t.entries, key)
	for i, have := range t.order {
		if have == key {
			t.order = append(t.order[:i], t.order[i+1:]...)
			break
		}
	}
}

// expired reports whether an entry's window has closed. The caller holds the lock.
func (t *Throttle) expired(e *attempts, now time.Time) bool {
	return t.Window <= 0 || now.Sub(e.first) >= t.Window
}

// evict makes room for one new entry, dropping the oldest (§19). The caller holds the lock.
//
// Oldest by insertion, which is also oldest by when its window opened. The alternative — dropping the
// least recently failed — would let an attacker cycling through names evict the very entry that is
// locking them out, since their own attempts keep the decoy names fresh.
func (t *Throttle) evict() {
	for len(t.entries) >= MaxThrottleKeys && len(t.order) > 0 {
		oldest := t.order[0]
		t.order = t.order[1:]
		delete(t.entries, oldest)
	}
}

// Tracked is how many names are counted against, for a test and for a diagnostic. It is a count and
// never the names (§19).
func (t *Throttle) Tracked() int {
	t.mu.Lock()
	defer t.mu.Unlock()
	return len(t.entries)
}
